### assignment_1 - task2

In [8]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
import time

In [2]:
#dtw function, window = 4
def dtw(s, t, window = 4):
    n, m = len(s), len(t)
    w = np.max([window, abs(n-m)]) # warping cannot be less than the difference in lengths. 
    dtw_matrix = np.zeros((n+1, m+1))
    
    for i in range(n+1):
        for j in range(m+1):
            dtw_matrix[i, j] = np.inf
    dtw_matrix[0, 0] = 0
    
    for i in range(1, n+1):
        for j in range(np.max([1, i-w]), np.min([m, i+w])+1):
            dtw_matrix[i, j] = 0
    
    for i in range(1, n+1):
        for j in range(np.max([1, i-w]), np.min([m, i+w])+1):
            cost = abs(s[i-1] - t[j-1])
            # take last min from a square box
            last_min = np.min([dtw_matrix[i-1, j], dtw_matrix[i, j-1], dtw_matrix[i-1, j-1]])
            dtw_matrix[i, j] = cost + last_min
    return dtw_matrix[-1,-1]

In [3]:
#import data
umd = pd.read_csv("./UMD_TEST.txt", sep=r"\s+",header = None)
umd

0         1         2         3         4         5         6    \
0    1.0  0.017644  0.030949  0.050555  0.044484  0.053277  0.041576   
1    1.0  0.041296  0.003551  0.027470  0.013158  0.009571  0.008074   
2    1.0 -0.000720  0.013283  0.029450  0.045201  0.006317  0.018805   
3    1.0  0.005201  0.013363  0.025733  0.026653  0.038946  0.012494   
4    1.0  0.022926  0.027036  0.011668  0.019500  0.036049 -0.001297   
..   ...       ...       ...       ...       ...       ...       ...   
139  3.0  0.008298  0.017240  0.010756  0.049488  0.005331  0.036829   
140  3.0  0.004801  0.042964  0.034519 -0.003426 -0.003353  0.005231   
141  3.0  0.033658  0.038235  0.009848  0.016061  0.026287  0.029979   
142  3.0  0.022805  0.020270  0.015546  0.038309  0.016303  0.011456   
143  3.0  0.012356  0.023970  0.013149  0.047396  0.024340  0.017960   

          7         8         9    ...       141       142       143  \
0    0.030947  0.027086  0.013764  ...  0.024575  0.033780  0.026589   
1    0.043743  0.040592  0.012190  ...  0.060539  0.046991  0.023586   
2    0.028901  0.013832  0.015240  ...  0.016442  0.039508  0.015171   
3    0.028303  0.032011  0.009467  ...  0.006383  0.037448  0.044335   
4    0.019717  0.039583  0.020628  ...  0.026997  0.036653  0.018117   
..        ...       ...       ...  ...       ...       ...       ...   
139  0.045530  0.046181  0.044288  ...  0.015062  0.028319  0.028443   
140  0.002309  0.029905  0.011177  ...  0.017705  0.045264  0.041434   
141  0.024099  0.023647  0.017566  ...  0.019306  0.022595  0.014942   
142  0.018286  0.015730  0.007319  ...  0.013016  0.030138  0.025736   
143  0.005501  0.040015 -0.000449  ...  0.014767  0.044482  0.000063   

          144       145       146       147       148       149       150  
0    0.013932  0.024928  0.022589  0.038248  0.049838  0.053419  0.040420  
1    0.001562 -0.002196  0.036730  0.039027  0.007754  0.004697  0.031440  
2    0.034708  0.010835  0.002942  0.006924  0.029502  0.040786  0.023144  
3    0.011143 -0.003624  0.001467  0.020991  0.027675  0.001621  0.015858  
4    0.018314  0.012536  0.040599  0.016590  0.032730  0.002498  0.011260  
..        ...       ...       ...       ...       ...       ...       ...  
139  0.051677  0.036134  0.038030  0.047388  0.006875  0.052691  0.051594  
140  0.040670  0.034613  0.001551  0.017716  0.001042  0.047787  0.020318  
141  0.009259  0.019186  0.035905  0.031095  0.006592  0.017948 -0.003614  
142  0.005955  0.050157  0.014435  0.005638  0.036126  0.010751 -0.005682  
143 -0.003054  0.037014  0.014849  0.031595  0.037902  0.008688  0.036395  

[144 rows x 151 columns]

## split feature_train, feature_test, label_train, label_test
I use the same window and value of parameter (test_size and random_state) for train_test_split to make sure result of this task can compare with that of task 1

In [4]:
#create label
label = []
feature = []
for i in range(len(umd)):
    label.append(umd.values[i][0])  # true label
    feature.append(umd.values[i][1:])

label = np.array(label)
feature = np.array(feature)

    #Dataset partitioning
    # parameters (feature value, target value); random_state is the random number seed
x_train, x_test, y_train, y_test = train_test_split(feature,label,test_size=0.33, random_state=42)
# x_train, x_test, y_train, y_test = train_test_split(feature,label)

print("y_test:\n", y_test)

y_test:
 [3. 1. 2. 3. 2. 1. 3. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 1. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]


## k-NN classifier
the knn_classifer function is a k-NN classifier. 
Parameter metric = dtw means this function use dtw function to calculate distance

In [5]:
def knn_classifer(x_train, x_test, y_train, y_test,k):
    
    # KNN estimator
    estimator = KNeighborsClassifier(n_neighbors = k, metric = dtw)
    estimator.fit(x_train, y_train) #Put in the feature values of the training set and the target values for training
    
    # model evaluation
    # method one: Direct comparison of true and predicted values
    y_predict = estimator.predict(x_test)
    print("y_predict:\n",y_predict)
    print("y_test:\n",y_test)
    print("comparing results:\n",y_test == y_predict)
    
    # method two: calculate accuracy
    score = estimator.score(x_test, y_test)
    print("accuray:\n",score)
    
    return None

In [10]:
# performance
start_time = time.time()
knn_classifer(x_train, x_test, y_train, y_test,3) #k = 3
end_time = time.time()
print("run_time:\n",end_time - start_time,"s")

y_predict:
 [3. 1. 2. 3. 2. 1. 2. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 1.
 2. 1. 1. 1. 1. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]
y_test:
 [3. 1. 2. 3. 2. 1. 3. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 1. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]
comparing results:
 [ True  True  True  True  True  True False  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True False
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True]
accuray:
 0.9583333333333334
run_time:
 147.41625118255615 s


## Verify that the 1-NN results
I set K = 1, and then compare this result with the result in task 1

result in task 1 is:
score of dtw: 0.9583333333333334 run_time is 90.58415198326111

result in here is:
accuray:0.9583333333333334 run_time: 140.84758806228638 s

conclusion: successfully verify that the 1-NN results are consistent

In [9]:
start_time = time.time()
print("comparing with 1-NN:")
knn_classifer(x_train, x_test, y_train, y_test,1)
end_time = time.time()
print("run_time:\n",end_time - start_time,"s")

comparing with 1-NN:
y_predict:
 [3. 1. 2. 3. 2. 1. 2. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 2. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]
y_test:
 [3. 1. 2. 3. 2. 1. 3. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 1. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]
comparing results:
 [ True  True  True  True  True  True False  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True False  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True]
accuray:
 0.9583333333333334
run_time:
 140.84758806228638 s


## k-NN - Euclidean ##
the only difference between knn_classifer_euclidean function and knn_classifer is that metric in knn_classifer is dtw 

when k = 3, the accuracy and run time of k-NN-Euclidean is 
accuray:
 0.9375
run_time:
 0.0189359188079834 s
 
while the accuracy and run time of k-NN-dtw is
accuray:
 0.9583333333333334
run_time:
 147.41625118255615 s
 
conclusion: the accurary of k-NN-dtw is better than that of euclidean

In [12]:
def knn_classifer_euclidean(x_train, x_test, y_train, y_test):
    
    #KNN estimator
    estimator = KNeighborsClassifier(n_neighbors = 3)
    estimator.fit(x_train, y_train) #Put in the feature values of the training set and the target values for training
    
    # model evaluation
    # method one: Direct comparison of true and predicted values
    y_predict = estimator.predict(x_test)
    print("y_predict:\n",y_predict)
    print("y_test:\n",y_test)
    print("comparing results:\n",y_test == y_predict)
    
    # method two: accuracy
    score = estimator.score(x_test, y_test)
    print("accuray:\n",score)
    
    return None

In [13]:
start_time = time.time()
knn_classifer_euclidean(x_train, x_test, y_train, y_test)
end_time = time.time()
print("run_time:\n",end_time - start_time,"s")

y_predict:
 [3. 1. 2. 3. 2. 1. 3. 2. 2. 1. 2. 2. 2. 2. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 2. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 2. 2. 1.]
y_test:
 [3. 1. 2. 3. 2. 1. 3. 2. 2. 1. 2. 2. 2. 3. 3. 2. 1. 2. 3. 2. 1. 3. 3. 3.
 2. 1. 1. 1. 1. 3. 1. 3. 1. 1. 1. 2. 3. 1. 3. 3. 3. 2. 1. 2. 1. 1. 2. 1.]
comparing results:
 [ True  True  True  True  True  True  True  True  True  True  True  True
  True False  True  True  True  True  True  True  True  True  True  True
  True  True  True  True False  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True False  True  True]
accuray:
 0.9375
run_time:
 0.0189359188079834 s
